opencloud: init at 1.1.0

christoph-heiss opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

[OpenCloud](https://opencloud.eu/) is a fully FLOSS and unencumbered fork of oCIS, backed by [OpenCloud GmbH](https://opencloud.eu/en/about-us/company) and [heinlein](https://www.heinlein-support.de/). See also their [annoucement](https://opencloud.eu/en/news/opencloud-available-now).

The packging properly builds it entirely from source. The service module & accompanying tests are mostly taken from oCIS, fixing a few things along the way.  

~~TODO:~~
- [x] Fix tests
- [x] Release notes entry

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

christoph-heiss committed:
opencloud: init at 1.1.0

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
opencloud-web: init at 2.0.0

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
opencloud-idp-web: init at 1.1.0

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5262

ShamrockLee commented on nixos/tests/all-tests.nix:
```diff
@@ -829,6 +829,7 @@ in {
   opensnitch = handleTest ./opensnitch.nix {};
   openssh = handleTest ./openssh.nix {};
   octoprint = handleTest ./octoprint.nix {};
+  opencloud = handleTest ./opencloud.nix {};
```
`runTest` seems preferred over `handleTest` per #386873.

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.md:
```diff
@@ -0,0 +1,110 @@
+sharing platform. It is a fork of oCIS, a ground-up rewrite of the well-known
+PHP-based NextCloud server.
+
+The server setup can be automated using
```
*can be configured

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.md:
```diff
@@ -0,0 +1,110 @@
+
+## Basic usage {#module-services-opencloud-basic-usage}
+
+OpenCloud is a golang application and does not require an HTTP server (such as
```
We always recommend nginx to get acme certs easily and this text kinda reads a bit like bad advice.

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.md:
```diff
@@ -0,0 +1,110 @@
+nginx) in front of it, though you may optionally use one if you will.
+
+OpenCloud is configured using a combination of yaml and environment
+variables. It is recommended to familiarize yourself with upstream's available
```
```suggestion
variables. You can find find the full documentation at <insert link here>.
```

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+      configDir = lib.mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        example = "/var/lib/opencloud/config";
```
This contradicts the above doc as there it was /etc/opencloud/config

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+
+          Note: This directory must contain at least a `opencloud.yaml`. Ensure
+          [user](#opt-services.opencloud.user) has read/write access to it. In some
+          circumstances you may need to add additional OpenCloud configuration files (e.g.,
```
In what circumstances? We should phrase it like: additional config files like proxy.yml have to be placed in this directory, too.

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+      user = lib.mkOption {
+        type = types.str;
+        default = defaultUser;
+        example = "yourUser";
```
This example is silly, just omit it

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+      group = lib.mkOption {
+        type = types.str;
+        default = defaultGroup;
+        example = "yourGroup";
```
Same here

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+      address = lib.mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = "Web interface address.";
```
*The address the interface will bind to.

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+      port = lib.mkOption {
+        type = types.port;
+        default = 9200;
+        description = "Web interface port.";
```
The port the web interface will listen on.

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+        type = types.str;
+        default = "https://localhost:9200";
+        example = "https://some-hostname-or-ip:9200";
+        description = "Web interface address.";
```
Same as above

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+
+      url = lib.mkOption {
+        type = types.str;
+        default = "https://localhost:9200";
```
How do the certs for localhost work? How are they trusted?

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+        description = ''
+          Extra config options.
+
+          Note that all the attributes here will be copied to /nix/store/ and will be world readable. Options like *_PASSWORD or *_SECRET should be part of     [environmentFile](#opt-services.opencloud.environmentFile) instead, and are only provided here for illustrative purpose.
```
This attributes are not copied, that is technically wrong. They are written to a file which is placed in the nix store.

Don't provide examples for illustration purposes if they are bad advice.

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+
+          Note that all the attributes here will be copied to /nix/store/ and will be world readable. Options like *_PASSWORD or *_SECRET should be part of     [environmentFile](#opt-services.opencloud.environmentFile) instead, and are only provided here for illustrative purpose.
+
+          Configuration here will override those from [environmentFile](#opt-services.opencloud.environmentFile) and will have highest precedence, at the cost of security. Do NOT put security sensitive stuff here.
```
Above this read the opposite

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+        example = {
+          OC_INSECURE = "false";
+          OC_LOG_LEVEL = "error";
+          OC_JWT_SECRET = "super_secret";
```
Drop this line and move it to the example of environmentFile

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+          User = cfg.user;
+          Group = cfg.group;
+          Restart = "always";
+          DynamicUser = true;
```
Dynamic user should not store large amounts of files as they are chownes on every start

SuperSandro2000 commented on nixos/modules/services/web-apps/opencloud.nix:
```diff
@@ -0,0 +1,212 @@
+    systemd.tmpfiles.settings."10-opencloud" = {
+      ${cfg.stateDir}.d = {
+        inherit (cfg) user group;
+        mode = "0700";
```
Shouldn't this be 750 to allow the group to read the files?

SuperSandro2000 commented on nixos/tests/opencloud.nix:
```diff
@@ -0,0 +1,226 @@
+          service_account_secret: .demKypQ$=pGl+yRar!#YaFjLYCr4YwE
+    '';
+
+    idmLdapCert = pkgs.runCommand "idm-ldap-cert" { buildInputs = [ pkgs.openssl ]; } ''
```
Use the snake oil certs from the acme test

SuperSandro2000 commented on pkgs/by-name/op/opencloud/idp-web.nix:
```diff
@@ -0,0 +1,60 @@
+  src = fetchFromGitHub {
+    owner = "opencloud-eu";
+    repo = "opencloud";
+    rev = "v${finalAttrs.version}";
```
This could be tag =

SuperSandro2000 commented on pkgs/by-name/op/opencloud/package.nix:
```diff
@@ -0,0 +1,119 @@
+in
+buildGoModule {
+  pname = "opencloud";
+  inherit version;
```
nit: there should be a new line after version

SuperSandro2000 commented on pkgs/by-name/op/opencloud/package.nix:
```diff
@@ -0,0 +1,119 @@
+  src = fetchFromGitHub {
+    owner = "opencloud-eu";
+    repo = "opencloud";
+    rev = "v${version}";
```
This should be tag

SuperSandro2000 commented on pkgs/by-name/op/opencloud/web.nix:
```diff
@@ -0,0 +1,54 @@
+  src = fetchFromGitHub {
+    owner = "opencloud-eu";
+    repo = "web";
+    rev = "v${finalAttrs.version}";
```
Should be tag

christoph-heiss committed:
nixos/opencloud: init module

Mostly taken from the existing oCIS module. Also cleaned up and properly
spell-checked.

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

dotlambda reviewed (commented):
Can we use `DynamicUser`?

FKouhai commented:
fwiw Im working on #401584 to package the desktop app for opencloud

K900 commented:
Would you be interested in me taking over the PR? I want to try running opencloud on my own infra.

K900 commented:
OK, let's continue this in https://github.com/NixOS/nixpkgs/pull/404891 then.
